### ISSUE TITLE
UserId too long to add Security Key WebauthN

### ISSUE THREAD

**Issue by jacac** [keycloak/keycloak#32143]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication/webauthn

### Describe the bug

Same issue as https://github.com/keycloak/keycloak/issues/24645. I could not reopen the other 

### Version

25.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

Ability to add security keys for federated users. Note that it may be needed to simulate this with LDAP provider with import disabled (To make sure that userId is in longer format like `f:<ldap provider UUID>:<ldap UUID>` or might be simulated with custom userStorage provider.

### Actual behavior

Error: Security key registration result is invalid. TypeError: User handle exceeds 64 bytes.

### How to Reproduce?

 Add `User federation` ldap should work well if the `username` is set to `userPrincipalName`. 

### Anything else?

I don't think that WebAuthn will change their standards which leaves it to Keycloak to adjust. Here is the link to the specification https://www.w3.org/TR/webauthn-2/#sctn-createCredential. It fails on point 5.

**Comment by mposolda** [comment]
Marking as important as the fact that federated users are not able to add WebAuthn keys is not ideal...

We may need the automated test of WebAuthn with LDAP (need to make sure to disable imports for LDAP to make sure that userIds are long format like `f:<ldap provider UUID>:<ldap UUID>` ). I don't know if it is feasible to test with custom userStorage provider or LDAP provider. We may need to doublecheck...

**Comment by kaweston** [comment]
My workplace is wanting to adopt WebAuthn and uses a custom User Storage provider; we have run into this very issue as we use UUIDs for the user id. Definitely a +1 to get this resolved at earliest opportunity.  

Can anyone kindly advise on the current status of this issue?

**Comment by jonkoops** [comment]
> Can anyone kindly advise on the current status of this issue?

If there have been no updates on the issue the status is the same. The issue is currently not assigned, which means nobody is working on it at this time. If you'd like to volunteer to fix it let us know.

**Comment by mposolda** [comment]
@rmartinc If you think that (1) does not have any impact on backwards compatibility, feel free to go with that. It could be an improvement at least.

Another option: Can we eventually use some user attribute to map the WebAuthn security-key to the user? Instead of using `userId`, we can do something like `user.setAttribute(someWebAuthnMappedAttributeFromTheWebAuthnPolicy, some-random-uuid)` . Then later during WebAuthn authentication, we can search for user by attribute and/or doublecheck if he has the attribute corresponding to the WebAuthn key. AFAIK we are allowing the same option also for the "X509 authentication" (The related class is `UserIdentityToModelMapper.UserIdentityToCustomAttributeMapper` . Could this work? We will also need to do this in backwards compatible way, so possibly allow to set this attribute in WebAuthn policy for new deployments (so new deployments work by default even for federated users), but have it empty when migrating from previous versions. WDYT?

**Comment by rmartinc** [comment]
@mposolda I think that your proposal would work but I don't know how useful it would be and if we are overcomplicating the configuration. Summarizing our options:

1.  We can add a new `KeycloakModelUtils.generateShortId()` that creates the UUID but it is encoded in base64url which transforms it in a 22 char ID instead of 36. The ldap component will use this short ID always (new ldap components will be created with the short ID and, in general, any user federation can use this type of ID). This gives 14 more chars for user IDs and, besides, using a custom Id is always an option for administrators.
2. The ldap uses the username as the internal ID. IMHO this is not OK because a rename of the user would change the ID in keycloak and, in turn, would make the webauthn association fail. I think we should add the ID from LDAP_ID (uuid) attribute if possible (I'm not sure if this is even feasible).
3. Your proposal. Extending the webauthn configuration to use an attribute as the `authn.id`. But my feeling here is that this is not going very usable. For example in the case of the ldap it would be the LDAP_ID attribute. But what happens with local DB users (there should be a fallback to the normal ID of the user if the attribute is missing, having the same issue again). What happens if a user without that attribute (it is using the internal ID until now) receives the attribute, the associations in webauthn are broken. My feeling is this is not really fixing the issue.

So my vote is for doing point 1 now, points 2 and 3 can be follow-ups if we see more work is needed on this subject.

**Comment by srowatt** [comment]
In our case we had the `externalId` as a UUID and the `providerId` was also the standard UUID so we ended up with a 77 char long federated user id `f:<PROVIDER_ID>:<USER_ID>` which is beyond the 64 webauthn id limit. We ended up taking a copy of `StorageId.java `(`FederatedStorageId.java`) and changing the `getId()` method to strip out the UUID hyphen chars and not include the colons or 'f' char suffix so the resulting id was 64 chars i.e. `<PROVIDER_ID><USER_ID>`. Then in the constructor we split the id into 2 x 32 char ids and add the hyphens back in so the `getProvider()` and `getExternalId()` methods still work. We then created a 

- custom `AbstractUserAdapterFederatedStorage` (`CustomUserStorage`) that used that new `FederatedStorageId.java` to manipulate the userId. 
- custom `UserStorageProvider` (`CustomUserStorageProvider`) that used the`CustomUserStorage` to load / find users
- custom `UserStorageProviderFactory` to create instances of the custom `CustomUserStorageProvider`
